Added tf.keras support

tgaddair opened the pull request:
Fixes #511

alsrgv commented on horovod/keras/__init__.py:
```diff
@@ -51,19 +52,7 @@ def get_gradients(self, loss, params):
-                return averaged_gradients
-        else:
-            return gradients
+        return _impl.get_gradients(gradients, self._device_dense, self._device_sparse, self._name)
```
I think we can move this whole `_DistributedOptimizer` implementation class as well as chunk of DistributedOptimizer in impl, and construct it within a function that implements DistributedOptimizer.

tgaddair replied:
Done.  Moved to `impl.py`, but note that it changes the module of the derived optimizer (see `test_keras.py` changes).

alsrgv commented on horovod/keras/callbacks.py:
```diff
@@ -24,17 +38,11 @@ def __init__(self, root_rank, device=''):
-            bcast_op = hvd.broadcast_global_variables(self.root_rank)
-            K.get_session().run(bcast_op)
+        _impl.BroadcastGlobalVariablesCallbackImpl.__init__(self, K, root_rank, device)
+        keras.callbacks.Callback.__init__(self)
```
Should we initialize superclass first?  https://stackoverflow.com/a/15777488

tgaddair replied:
It's multiple inheritance, so both are superclasses.  But using this "old style" multiple inheritance was kinda ugly and error prone, so I switched it to "cooperative multiple inheritance".

alsrgv commented on horovod/keras/callbacks_impl.py:
```diff
@@ -0,0 +1,177 @@
+class LearningRateWarmupCallbackImpl(LearningRateScheduleCallbackImpl):
+    def __init__(self, backend, warmup_epochs=5, momentum_correction=True, steps_per_epoch=None,
+                 verbose=0):
+        """
```
Since you did not copy-paste other comments, we can nuke this one as well.

tgaddair replied:
Done.

tgaddair commented:
Fixed unit tests, though the nightly tests are still failing due to PyTorch installation issues.  I added backwards compatibility with older versions of TensorFlow.

tgaddair commented on horovod/tensorflow/keras/__init__.py:
```diff
@@ -0,0 +1,155 @@
+# limitations under the License.
+# ==============================================================================
+
+# Copyright 2017 Uber Technologies, Inc. All Rights Reserved.
```
Oops!  Fixed.
